File: Infrastructure/Mappers/ProductMapper.cs
using Domain.Contracts.Products;
using Domain.Models;
using Infrastructure.DbEntities;

namespace Infrastructure.Mappers;

public static class ProductMapper
{
    public static ProductDbEntity FromDomainToDbEntity(Product domain)
    {
        return new ProductDbEntity(
            id: domain.Id.Value, 
            dateCreated: domain.DateCreated, 
            name: domain.Name,
            description: domain.Description,
            price: domain.Price.Value,
            amount: domain.Amount.Value
        );
    }

    public static Product FromDbEntityToDomain(ProductDbEntity dbEntity)
    {
        var contract = new CreateProductContract(
            id: dbEntity.Id, 
            name: dbEntity.Name,
            price: dbEntity.Price,
            description: dbEntity.Description,
            dateCreated: dbEntity.DateCreated,
            images: dbEntity.Images.Select(ProductImageMapper.ToDomain).ToList(),
            amount: dbEntity.Amount
        );

        return Product.ExecuteCreate(contract);
    }
}

File: frontend/src/infrastructure/mappers/productMapper.ts
import IProduct from "../../domain/models/IProduct";
import IProductApiModel from "../apiModels/IProductApiModel";
import ImageDataMapper from "./productImageMapper";

const productMapper = {
    apiToDomain: (source: IProductApiModel): IProduct => {
        return {
            id: source.id,
            name: source.name,
            price: source.price,
            description: source.description,
            dateCreated: new Date(source.dateCreated),
            images: source.images.map(ImageDataMapper.apiToDomain),
            amount: source.amount
        };
    },
};

export default productMapper;
